File: 209.Minimum-Size-Subarray-Sum.js
/**
 * @param {number} target
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function (target, nums) {
  let size = Infinity;
  let sum = 0;
  let left = 0;
  let right = 0;

  while (left < nums.length) {
    if (sum < target) {
      if (right === nums.length) break;

      sum += nums[right];
      right += 1;
    } else {
      const newSize = right - left;
      size = Math.min(size, newSize);

      sum -= nums[left];
      left += 1;
    }
  }

  return size === Infinity ? 0 : size;
};

File: python/209.Minimum-Size-Subarray-Sum.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        size = float("inf")
        left = 0
        right = 0
        total = 0

        while left < len(nums):
            if total < target:
                if right == len(nums):
                    break

                total += nums[right]
                right += 1
            else:
                new_size = right - left
                size = min(size, new_size)
                
                total -= nums[left]
                left += 1

        return 0 if size == float("inf") else size